Sweetener.Logging/LogPattern.cs

```diff
@@ -0,0 +1,385 @@
+﻿using System;
+using System.Diagnostics;
+using System.Text;
+using System.Threading;
+
+namespace Sweetener.Logging
+{
+    /// <summary>
+    /// Encapsulates the logic for parsing and formatting log entry patterns.
+    /// </summary>
+    internal class LogPattern
+    {
+        /// <summary>
+        /// The internal composite format string used by the <see cref="LogPattern"/>.
+        /// </summary>
+        internal readonly string CompositeFormatString;
+
+        /// <summary>
+        /// The default log layout pattern.
+        /// </summary>
+        public const string Default = "[{ts:O}] [{level:F}] {msg}";
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="LogPattern"/> class based on
+        /// a string layout of each log entry.
+        /// </summary>
+        /// <param name="pattern">A format string that describes the layout of each log entry.</param>
+        /// <exception cref="ArgumentNullException"><paramref name="pattern"/> is <c>null</c>.</exception>
+        /// <exception cref="FormatException">The <paramref name="pattern"/> is not formatted correctly.</exception>
+        public LogPattern(string pattern)
+        {
+            CompositeFormatString = Resolve(pattern);
+        }
+
+        /// <summary>
+        /// A delegate that yields a log entry using a pre-defined composite string.
+        /// </summary>
+        /// <param name="provider">An <see cref="IFormatProvider"/> object for a specific culture.</param>
+        /// <param name="level">The level associated with the log entry.</param>
+        /// <param name="message">The value requested for logging.</param>
+        /// <returns>The resolved log entry based on the pattern.</returns>
+        public string Format(IFormatProvider provider, LogLevel level, string message)
+            => Format(provider, DateTime.UtcNow, level, message);
+
+        /// <summary>
+        /// A delegate that yields a log entry using a pre-defined composite string.
+        /// </summary>
+        /// <param name="provider">An <see cref="IFormatProvider"/> object for a specific culture.</param>
+        /// <param name="timestamp">The time associated with the log entry.</param>
+        /// <param name="level">The level associated with the log entry.</param>
+        /// <param name="message">The value requested for logging.</param>
+        /// <returns>The resolved log entry based on the pattern.</returns>
+        internal string Format(IFormatProvider provider, DateTime timestamp, LogLevel level, string message)
```

Comment:
At some point in the near future, I can add logic to emit a more optimized function if only a subset of the parameters are used.